[DAG] Fold add(mul(add(A, CA), CM), CB) -> add(mul(A, CM), CM*CA+CB)

davemgreen opened the pull request:
This is useful when the inner add has multiple uses, and so cannot be
canonicalized by pushing the constants down through the mul. I have added
patterns for both `add(mul(add(A, CA), CM), CB)` and with an extra add
`add(add(mul(add(A, CA), CM), B) CB)` as the second can come up when lowering
geps.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -2838,6 +2838,36 @@ SDValue DAGCombiner::visitADDLike(SDNode *N) {
+      SDValue Mul =
+          DAG.getNode(ISD::MUL, SDLoc(N1), VT, A, DAG.getConstant(CM, DL, VT));
+      return DAG.getNode(
+          ISD::ADD, DL, VT, Mul,
```
Can flags be preserved?

arsenm replied:
I mean I always just stick it in alive and see if it works

davemgreen replied:
This is the generic combine: https://alive2.llvm.org/ce/z/k7Yvo3. In general the flags need removing.

In specific cases the flags might be retained, but it feels to me a bit niche and difficult to specify: https://alive2.llvm.org/ce/z/aK5Az3.

Let me know what you think.

arsenm replied:
I think using constants in the proof is making this more permissive looking than it should be

davemgreen replied:
What do you mean, sorry? That it is OK without the flags?

arsenm replied:
I mean the proof you wrote here isn't useful with small constants. You would ideally rewrite it with variables, and then assume calls if you need to restrict the ranges in some way

davemgreen replied:
The first link is the general proof with any values. noundef %CA/%CB/%CM are the constants. In general the flags from the add/mul need to be dropped.
The second link was a specific example where certain constant can keep nuw/nsw, but it is specific to those constants.

It does look like if all the input adds/mul are nsw/nuw then we might be able to keep some flags on the remaining instructions. I'm not sure if it's worth it considering all the other transforms in DAG, but I can try and add that to the patch.

arsenm replied:
I played around with a bit and I'm not sure what the rule is, probably best to keep that in a separate patch if it's worth it

davemgreen replied:
Oh - sorry. This is the version with all nuw: https://alive2.llvm.org/ce/z/nxaMNR
And the version with both: https://alive2.llvm.org/ce/z/65Xyyf
It didn't apply with just nsw though: https://alive2.llvm.org/ce/z/7eZnnp

I can remove that patch if you like, it would be just a case of dropping the last patch from this review.

arsenm replied:
Keep it - as long as some tests actually hit this

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -2838,6 +2838,36 @@ SDValue DAGCombiner::visitADDLike(SDNode *N) {
+    if (sd_match(N0, m_OneUse(m_Mul(m_Add(m_Value(A), m_ConstInt(CA)),
+                                    m_ConstInt(CM)))) &&
+        TLI.isLegalAddImmediate(
+            (CA * CM + CB->getAPIntValue()).getSExtValue())) {
```
This will fail for any APInt size > 64-bits

davemgreen replied:
Ah, yes. Good point. I'll add a fix now.

davemgreen commented on llvm/test/CodeGen/AArch64/addimm-mulimm.ll:
```diff
@@ -140,11 +140,208 @@ define signext i32 @addimm_mulimm_reject_02(i32 signext %a) {
+else:
+  ret i32 0
+}
+
```
OK - will do.

davemgreen committed:
[DAG] Fold add(mul(add(A, CA), CM), CB) -> add(mul(A, CM), CM*CA+CB)

This is useful when the inner add has multiple uses, and so cannot be
canonicalized by pushing the constants down through the mul. I have added
patterns for both `add(mul(add(A, CA), CM), CB)` and with an extra add
`add(add(mul(add(A, CA), CM), B) CB)` as the second can come up when lowering
geps.
